### PR TITLE
Fix in audit_rules_systadmin_actions and new rule audit_rules_sysadmi…

### PR DESCRIPTION
**Pull request by rumch-se**
…n_scope

#### Description:

- _Fixes in ansible part. Creation of a new rule_

#### Rationale:

- The current rule includes 2 CIS recommendations. With this PR a new rule was created to cover the recommendation related to scope

### CODE REVIEW

**Comment by rumch-se** [comment]
Hello @vojtapolasek 
I think that macros created by you ansible_audit_augenrules_add_watch_rule, and ansible_audit_auditctl_add_watch_rule do not work as expected in a specific situation like this.
1/
On SUSE we have to add 4 records for auditing in audit.rules located in /etc/audit or to created 2 files into /etc/audit/rules.d/ - scope.rules and privileged-actions.rules which include these records

these records are: 
-w /etc/sudoers -p wa -k actions
-w /etc/sudoers.d -p wa -k actions
-w /etc/sudoers -p wa -k scope
-w /etc/sudoers.d -p wa -k scope
2/When I checked for existence of the actions with  oscap xccdf eval command - my test failed because I did not have records and file - this was expected result
3/I executed ansible remediation - and my system was fixed - this was expected result
4/I checked for the existence of the scope via oscap xccdf eval command - my test failed because I did not have records and file for the keyword scope - this was expected result
5/I executed ansible remediation - and my system was not fixed 

You can simulate this manually 
add these records based on keyword scope
-w /etc/sudoers -p wa -k scope
-w /etc/sudoers.d -p wa -k scope
to /etc/audit/audit.rules 
and
create a file scope.rules in /etc/audit/rules.d/ which include these records

Now execute  oscap xccdf eval command which checks for the rule audit_rules_sysadmin_actions - it should fail when you don't have expected file and records
Now execute a command for ansible remediation againts the rule audit_rules_sysadmin_actions and to see if your system will be remediated.
I think that NO, and the problem is coming from the fact that in your macro you check only the beginning of the line: -w /etc/sudoers -p wa, but you don't check for the remaining part with the key. 

Because of that when we have similar rules which begin with same string only the 1st rule will make remediation (will add rules), but the remaining rules will not. For example I tried this - to check/remediate my system with the keyword scope - it worked, but after I was not able to remediated my system with the keyword actions. 

NOTE: ALL TESTS WHICH I DID WERE IN ANSIBLE, I DID NOT TEST BASH. 


Have a nice day
Rumen

File: linux_os/guide/system/auditing/auditd_configure_rules/audit_rules_sysadmin_actions/ansible/shared.yml
Comment (teacup-on-rockingchair) [inline linux_os/guide/system/auditing/auditd_configure_rules/audit_rules_sysadmin_actions/ansible/shared.yml:1]: @rumch-se I am not sure what happens here but do you get different results if you change the order of the macros, because otherwise the files are identical right?

**Comment by teacup-on-rockingchair** [comment]
> 

If I understand correctly the patches in the PR do not address the issue right?

As far as I could see an example of the problem is that if one is applied remediation for the `audit_rules_sysadmin_actions` rule for example, later on the `audit_rules_sysadmin_scope` remediation will not work, since the key param is not checked by the bash and ansible macros as part of the audit rules statements.

Also can confirm that the above is valid for bash remediation also.

**Review by teacup-on-rockingchair — CHANGES_REQUESTED** [review]

File: linux_os/guide/system/auditing/auditd_configure_rules/audit_rules_sysadmin_actions/ansible/shared.yml
Comment (rumch-se) [reply linux_os/guide/system/auditing/auditd_configure_rules/audit_rules_sysadmin_actions/ansible/shared.yml:1]: Hello @teacup-on-rockingchair,

No I did not get different results. I changed the order of macros in ansible, to correspond to the order of macros in bash. But the main problem is in macros, because if we have 2+ rules which have identical parts - only the 1st executed rule will remediate the system.  At the same time documentation of the macros shows that we have to be able to add new rows - which is not correct statement. This is the reason why I have addressed the issue to the creator of the macros.

Have a nice day.
Rumen

**Comment by vojtapolasek** [comment]
Hello @rumch-se, I think the Ansible macros are working correctly.
If I understand your question correctly, your goal is to add two pairs of rules into two different files in /etc/audit/rules.d directory. The rules are nearly identical, the only part which is different is their key.
But macros disregard the key because it does not affect the behavior of the rule - the only purpose of the key is to make the log searching easier. Therefore, only one instance of those rules will be present. The key parameter is even not checked in OVAL.
Honestly, I am not sure what is the purpose of adding those rules twice. I am afraid that you will actually don't get the expected result, you can try it your self:
```
touch /root/test
echo "-w /root/test -p wa -k hello" > /etc/audit/rules.d/test.rules
echo "-w /root/test -p wa -k world" >> /etc/audit/rules.d/test.rules
augenrules --load
echo "testing" > /root/test
ausearch -k hello
ausearch -k world
```
I got different results for each key, although the rules were the same.
I checked latest CIS benchmarks for SLE 15 and SLE 12 and I see there the rule ending with the "scope" key. But the rule checking for "actions" key actually has different watch part - it checks for the Sudo log file, not for the Sudo configuration.
In conclusion, trying to remediate two Audit rules where the only difference is the key is something which should not happen.

**Comment by rumch-se** [comment]
Hello @vojtapolasek 

There are two rules which are nearly identical, and this macro does not manage them well. The pictures are attached. Also @teacup-on-rockingchair can confirm that bash part also does not work as expected. The problem is that you don't check the whole string but only part of it. The expected behavior should be when the audit rule is not exist it should be added, when exist - do nothing. But when you check for the existence you check only the first part of the audit rule, and when they are equal, but key is different - the audit rules with different key will not be added. Please see the pictures and simulate a case when you have already one audit rule and try to add another rule but with a different key.  
Note: In case of CIS - adding rules for actions and scope will work, but DISA has a different view about actions .. and in this case adding of rules for action and scope will be not possible. 
macro-1
macro-2

**Comment by vojtapolasek** [comment]
Hello @rumch-se. Thank you for the screenshots. I still think that the way in which the macro works now is reasonable, and mainly that rules audit_rules_sysadmin_scope and audit_rules_sysadmin_actions should NOT be in the same profile.
I understand the problem you are facing, but I think that the problem is not due to macros but to incorrect composition of rules.
In which profile do you face the problem you describe?

Please see rationale and discussion in this PR: https://github.com/ComplianceAsCode/content/pull/9463
Moreover, I still think that utilizing two audit rules which watch for the same file and same permissions and the only difference is the key is bad practice. @stevegrubb, could you please confirm what is the result of of using two Audit watch rules watching the same path and same permission changes with only difference being the key? My understanding is that only one of those rules wil get processed or results will not be consistent.

**Comment by rumch-se** [comment]
Hello @vojtapolasek, @stevegrubb 

I am not agree with you @vojtapolasek, because now there is a risk of an "open door", if I modify the file /etc/audit/audit.rules  by using this rule: -w /etc/sudoers -p wa -k skip  and have an actions.file in /etc/audit/rules.d/ with same rules - the oval check will pass, and remediation will not add a record .i.e a potential  attacker can modify audit part of the system in his favor, and like that the automation will be broken. In the information security we have the principle - "Expect the unexpected" and we as security experts want to control the situation via hardening. You can see the screenshot - which shows that. 

Have a nice day
Rumen
        
macro_3

**Comment by teacup-on-rockingchair** [comment]
> ausearch -k world

@vojtapolasek  if I understand correctly your point is that, the audit rules main point is to watch access to files, and the oval should be following that.

At the same time `-k somekey` is just convenience for searching/deleting the auditing rules.

If we are following those guidance the new rule oval will be better off just checking the main part of the rule , instead of the key and the same probably should be valid of remediations also?

**Comment by vojtapolasek** [comment]
@teacup-on-rockingchair  yes, my point is that the key specified after the -k argument is there only for humans. It does not affect the event which the rule is looking for.
It is explained here: https://www.man7.org/linux/man-pages/man7/audit.rules.7.html#NOTES
My point is that our automation should NOT enforce usage of specific keys - this is up to the system administrator.

**Comment by rumch-se** [comment]
Hello @vojtapolasek

I would like to clarify point here.

If somebody in moment t0 succeeded to include -w /etc/sudoers -p wa -k skip, then the administrator via automation will will not include a rule with his desired key - for example action in a moment t1 - because the audit will show pass. 
In this case when the logs should be analyses by the administrator in moment tn - he will use an automatic log system, or he will use grep with the key defined by him, and he will not find any log records, because these records will have key skip instead of correct key action. In other words the activities of the attacker will be obfuscated. 

Have a nice day
Rumen

**Comment by teacup-on-rockingchair** [comment]
> @teacup-on-rockingchair yes, my point is that the key specified after the -k argument is there only for humans. It does not affect the event which the rule is looking for. It is explained here: https://www.man7.org/linux/man-pages/man7/audit.rules.7.html#NOTES My point is that our automation should NOT enforce usage of specific keys - this is up to the system administrator.

I tend to agree with @rumch-se  with that the scenario with the `skip` keys maybe a valid one, but also due to the cited notes in the manual we obviously cannot enforce that policy to the common framework.
So I would suggest to SLES specific solution for this CIS rule, which will both enforce the key as integral part of the oval check and remediations and also will not affect other platforms. Of course the only possible problem would be executing  audit_rules_systadmin_actions and new rule audit_rules_sysadmin_scope one after another, but as long as we keep those rules in separate profiles, this should not be an issue. I am guessing somebody from @ComplianceAsCode/content-development can also give opinon if that looks as workable solution?

**Comment by vojtapolasek** [comment]
Hello @rumch-se  @teacup-on-rockingchair, thank you for explaining your point of view.
I am glad that we all agree that putting both rules in one profile is not a good idea.

> If somebody in moment t0 succeeded to include -w /etc/sudoers -p wa -k skip, then the administrator via automation wil not include a rule with his desired key - for example action in a moment t1 - because the audit will show pass.

I think that if an attacker manages to write into that file which is only writable by root, you have much bigger problem than missing Audit records.

> In this case when the logs should be analyses by the administrator in moment tn - he will use an automatic log system, or he will use grep with the key defined by him, and he will not find any log records, because these records will have key skip instead of correct key action. In other words the activities of the attacker will be obfuscated.

The key is not the only parameter which you can use for searching in Audit logs.

I still think that having such a rule can bring more problems than it is worth it, but I respect your approach.
However, I don't think the change you suggest should happen for ALL audit rules in the project. I can imagine that the macro can get an additional parameter which will indicate if the remediation will check for the exact line or not.
Feel free to extend the macro along those lines.
Also please mention the fact that you are checking for the exact line either in the rule title or in the rule description.
I hope this is acceptable for all of us.

**Comment by marcusburghardt** [comment]
Regarding the discussion about the `key` value, the value informed as `key` is just used for convenience. It has no technical impact in the other parameters of the rule. If the admin informs anything or nothing, the events are still collected according to the other parameters in the rule. So I agree these macros are correct for not caring about the key. The value entered as a key or even the use of keys is at the discretion of the administrator.

Furthermore, relying exclusively on keys to search logs seems to be a weak approach in the case of a real audit.
The requirement asks to ensure that events are collected, but does not (and should not) dictate how an auditor must investigate these logs.

I would be fine if you extend the macro to **optionally** consider keys in audit rules since this does not affect the current behavior. I would also respect if you desire to enforce specific keys in SUSE, but I would recommend to reconsider this approach because I can imagine it bringing more issues than value to the users.

**Comment by rumch-se** [comment]
Hello @marcusburghardt

I understand that the key is used for convenience, and its definition is in discretion of administrator - i.e. he/she has to control the usage of the key. We will have a problem when the administrator has to manage a lot of systems - in my practice we had an administrator responsible for 300+ servers on daily basis - and he had to use automation tools to succeeded with his tasks. According to the current realization of the oval check  (which we use for audit) - the test will pass when we have any key after the -k option. There is a risk the administrator of the system not to know that for a specific type of event the key is not equal to the key he expected to be used. Usually after each patch (security update) - the administrator have to execute again audit+remediation. If the audit checks also for the key, the administrator will know that something is not OK with a specific system.  From a security point of view we have to control input and not to allow something to be injected - otherwise we will have impact on the results ... and the results are important for the next actions. In the bank industry there is a classical example - when the programmer of the banking system used the exchange rate to his benefits - he transferred from customer accounts to his account the value after the 2nd position of the decimal point.  (when the exchange rate is 1.23465 - he kept 1.23 in the customer account and took the remaining part). From such of rounding operations "he made" ~1000 daily.   

Have a nice day
Rumen

**Comment by marcusburghardt** [comment]
> Hello @marcusburghardt
> 
> I understand that the key is used for convenience, and its definition is in discretion of administrator - i.e. he/she has to control the usage of the key. We will have a problem when the administrator has to manage a lot of systems - in my practice we had an administrator responsible for 300+ servers on daily basis - and he had to use automation tools to succeeded with his tasks. According to the current realization of the oval check (which we use for audit) - the test will pass when we have any key after the -k option. There is a risk the administrator of the system not to know that for a specific type of event the key is not equal to the key he expected to be used. Usually after each patch (security update) - the administrator have to execute again audit+remediation. If the audit checks also for the key, the administrator will know that something is not OK with a specific system. From a security point of view we have to control input and not to allow something to be injected - otherwise we will have impact on the results ... and the results are important for the next actions. In the bank industry there is a classical example - when the programmer of the banking system used the exchange rate to his benefits - he transferred from customer accounts to his account the value after the 2nd position of the decimal point. (when the exchange rate is 1.23465 - he kept 1.23 in the customer account and took the remaining part). From such of rounding operations "he made" ~1000 daily.
> 
> Have a nice day Rumen

This administrator can check the same events regardless of the keys values. But as I said, I understand if you prefer to enforce a specific key for audit rules, but please don't change this behavior for all rules. You can make it optional and not enabled by default.

**Comment by stevegrubb** [comment]
Late to the game...
1) having 2 rules to the same file with different key is a waste on memory. The first rule will always win. However, if the watch is on a directory and one has the trailing slash and the other doesn't, they are different rules. One watches the files in the directory and the other watches the directory entries itself to see if anything gets added or deleted. It is for this reason that '-w' should be avoided and '-F path' and '-F dir' should be used to make this very clear. -F path on a directory watches the directory entries. -F dir watches the files in the directory.
2) "-k argument" is there to aid in investigations. Aureport --summary --key can help see what has been happening on the system. Then you can see what users are triggering the rule, ausearch --key some-key --raw | aurport --summary --user -i. It can also be used to tie several rules together that work together looking for a specific problem.
3) The audit package tries to ship rules with opinionated keys to meet specific requirements. You can always add a second or third key to any syscall rule.

**Comment by rumch-se** [comment]
Hello @marcusburghardt and @teacup-on-rockingchair 

Would be possible this PR to be merged to the master? 
Have a nice day
Rumen

File: controls/cis_sle15.yml [hunk 1]
```diff
@@ -1469,7 +1469,9 @@ controls:
     levels:
       - l2_server
       - l2_workstation
-    status: manual # missing rule for sudoers scope
+    status: automated
```
Comment (marcusburghardt) [inline controls/cis_sle15.yml:1472]:
```suggestion
    status: automated
```

File: linux_os/guide/system/auditing/auditd_configure_rules/audit_rules_sysadmin_scope/rule.yml [hunk 2]
```diff
@@ -0,0 +1,62 @@
+documentation_complete: true
+
+prodtype: sle12,sle15
+
+title: 'Ensure auditd Collects System Administrator Scope'
+
+description: |-
+    Monitor changes in scope for system administrators, when the system is
```
Comment (marcusburghardt) [inline linux_os/guide/system/auditing/auditd_configure_rules/audit_rules_sysadmin_scope/rule.yml:8]:
```suggestion
    Monitor changes in scope for system administrators, when the system is
```

File: linux_os/guide/system/auditing/auditd_configure_rules/audit_rules_sysadmin_scope/rule.yml [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    <pre>-w /etc/sudoers -p wa -k scope
+    -w /etc/sudoers.d/ -p wa -k scope</pre>
+    If the <tt>auditd</tt> daemon is configured to use the <tt>auditctl</tt>
+    utility to read audit rules during daemon startup, add the following line to
+    <tt>/etc/audit/audit.rules</tt> file:
+    <pre>-w /etc/sudoers -p wa -k scope
+    -w /etc/sudoers.d/ -p wa -k scope</pre>
+
+rationale: |-
+    Changes in the /etc/sudoers file, or a file in the /etc/sudoers.d/ directory
```
Comment (marcusburghardt) [inline linux_os/guide/system/auditing/auditd_configure_rules/audit_rules_sysadmin_scope/rule.yml:24]:
```suggestion
    Changes in the /etc/sudoers file, or a file in the /etc/sudoers.d/ directory
```

File: linux_os/guide/system/auditing/auditd_configure_rules/audit_rules_sysadmin_scope/rule.yml [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    -w /etc/sudoers.d/ -p wa -k scope</pre>
+    If the <tt>auditd</tt> daemon is configured to use the <tt>auditctl</tt>
+    utility to read audit rules during daemon startup, add the following line to
+    <tt>/etc/audit/audit.rules</tt> file:
+    <pre>-w /etc/sudoers -p wa -k scope
+    -w /etc/sudoers.d/ -p wa -k scope</pre>
+
+rationale: |-
+    Changes in the /etc/sudoers file, or a file in the /etc/sudoers.d/ directory 
+    can indicate that an unauthorized change has been made to scope of system
```
Comment (marcusburghardt) [inline linux_os/guide/system/auditing/auditd_configure_rules/audit_rules_sysadmin_scope/rule.yml:25]:
```suggestion
    can indicate that an unauthorized change has been made to scope of system
```

**Review by marcusburghardt — CHANGES_REQUESTED** [review]
This PR is basically duplicating the `audit_rules_sysadmin_actions` rule without adding value to the project. The `audit_rules_sysadmin_scope` rule is unnecessary and including both `audit_rules_sysadmin_scope` and `audit_rules_sysadmin_actions` in the same profile, as tried in this PR, will cause avoidable waste of resources and confusion.

If you want to create a mechanism where the `key` is assessed, you should implement this mechanism optionally in the existing OVAL. You can also introduce a variable if you want more flexibility for keys names in remediation.

Finally, I strongly recommend to avoid creating duplicated audit rules with different keys. If the Benchmark is asking this, the wiser approach would be to update the benchmark instead. As mentioned by the author of the `auditd`, it is a waste of memory with no real benefit because the first ~~processed~~ matching rule will always win.

**Comment by marcusburghardt** [comment]
> Hello @marcusburghardt and @teacup-on-rockingchair
> 
> Would be possible this PR to be merged to the master? Have a nice day Rumen

I don't think so @rumch-se . There are serious concerns about this approach, as mentioned in the comments. They should first be solved.

**Comment by rumch-se** [comment]
Hello @marcusburghardt 

I have considered your last proposals and 
 - removed from the PR the rule audit_rules_sysadmin_scope
 - adapted the rule audit_sudo_log_events

After these changes the rule  audit_rules_sysadmin_actions covers SLE12/15 CIS requirement 4.1.14 - Ensure changes to system administration scope (sudoers) is collected (Automated), and the rule  audit_sudo_log_events covers SLE12/15 CIS requirement Ensure system administrator actions (sudolog) are collected (Automated)

Have a nice day
Rumen

**Comment by marcusburghardt** [comment]
> Hello @marcusburghardt
> 
> I have considered your last proposals and
> 
> * removed from the PR the rule audit_rules_sysadmin_scope
> * adapted the rule audit_sudo_log_events
> 
> After these changes the rule audit_rules_sysadmin_actions covers SLE12/15 CIS requirement 4.1.14 - Ensure changes to system administration scope (sudoers) is collected (Automated), and the rule audit_sudo_log_events covers SLE12/15 CIS requirement Ensure system administrator actions (sudolog) are collected (Automated)
> 
> Have a nice day Rumen

Thanks for considering @rumch-se . The concerns were solved with this last commit. I will wait the CI tests to finish but it seems good to me.

**Comment by stevegrubb** [comment]
"first processed rule will always win"

That should be first **matching** rule wins.

**Review by marcusburghardt — APPROVED** [review]
LGTM. Thanks @rumch-se .
It is up to you now @teacup-on-rockingchair

**Review by teacup-on-rockingchair — APPROVED** [review]
:+1:
